Update odhargo component to v2.12.5

tumido opened the pull request:
**Merge together with https://github.com/opendatahub-io/odh-dashboard/pull/37 please**

Argo is moving forward quite fast and ODH should keep up with the progress. This PR implements 2 enhancements to the current Argo in ODH:

- Bump the manifests to the 2.12.5 release
- Manifests are generated via #289 script

That required me to change how the manifest files are structured in odhargo though - all custom ODH manifests and modifications are now static, while all the resources which are built via kustomize from the upstream are built into specific files:

For `cluster/base` those files are:
- `crds.yaml`
- `clusterroles.yaml`

For `odhargo/base` it's a single file: `namespace-install.yaml` that is very similar to the upstream's [`namespace-install.yaml`](https://github.com/argoproj/argo/blob/master/manifests/enamespace-install.yaml) but it lacks CRDs.

Target version used by the generator script is the latest release, fetched during the script execution.

One last modification to the previous behavior is that Argo can now use the `default` service account (which is used by default by argo anyways) instead of creating a custom one in and for ODH only.

tumido commented:
/retest

tumido commented:
/retest

nakfour commented:
@nakfour  to review

tumido commented on odhargo/odhargo/base/argo-server-route.yaml:
```diff
@@ -1,11 +1,10 @@
 kind: Route
 apiVersion: route.openshift.io/v1
 metadata:
-  name: argo-portal
```
This may be a potential issue and would need remapping in odh-dashboard. I've renamed the `argo-portal` route to `argo-server` since both the [`Service`](https://github.com/opendatahub-io/odh-manifests/pull/208/files#diff-5fe25086066c4f4201a497a261b3214ce38a12f9ac1ae29a4669b5f729d8b838R201) and [`Deployment`](https://github.com/opendatahub-io/odh-manifests/pull/208/files#diff-5fe25086066c4f4201a497a261b3214ce38a12f9ac1ae29a4669b5f729d8b838R226) in the upstream manifests (as well as in the renders here) are using the `argo-server` name, so we're rather consistent.

tumido replied:
Resolved via https://github.com/opendatahub-io/odh-dashboard/pull/37

tumido commented:
Let's wait a bit for v2.12.0. It's already rc3 as of today. I'll run the updater again once the final release is available.

LaVLaS commented on odhargo/README.md:
````diff
@@ -1,37 +1,49 @@
 ```
 
 ### Run a Workflow
-To run an example workflow, you can use the portal UI to run the example workflow, however the `serviceAccountName` has to be added to the workflow as shown below. We created this example `serviceAccount` as part of the installation to give basic permissions to a workflow. You can edit this `serviceAccount` and add more pemissions if your workflow requires that.
````
Does this updated version of Argo remove the need a service account with elevated privileges to allow for advanced Workflow features (artifact passing) that fail(ed) with the `default` serviceAccount?

tumido replied:
When it comes to artifact passing access to pods and logs is all Argo needs. So with just these permissions Argo Workflows should work just fine.

We should probably link the upstream docs in here though:
- https://argoproj.github.io/argo/service-accounts/
- https://argoproj.github.io/argo/workflow-rbac/

Mind that Argo uses service accounts to enforce proper RBAC on workflows (different SA is used to execute the workflow for different user roles), so ODH should not be much opinionated here.
- https://argoproj.github.io/argo/argo-server-sso/#sso-rbac

LaVLaS replied:
Documenting this from an offline conversation, we want components to be standalone and preserve the default permissions for the `default` ServiceAccount.

tumido replied:
I've reintroduced the `argo-workflow` SA, removed all changes to the `default` SA and assinged `argo-workflow` SA as the default for Workflow execution. WDYT? :slightly_smiling_face:

LaVLaS commented on odhargo/README.md:
````diff
@@ -1,37 +1,49 @@
 ### Run a Workflow
-To run an example workflow, you can use the portal UI to run the example workflow, however the `serviceAccountName` has to be added to the workflow as shown below. We created this example `serviceAccount` as part of the installation to give basic permissions to a workflow. You can edit this `serviceAccount` and add more pemissions if your workflow requires that.
-```
-apiVersion: argoproj.io/v1alpha1
````
I think we should keep these instructions for manual creation of `hello-world` workflow but reference that the argo cli can be used to managed workflows from the command line.

tumido commented:
This is not WIP anymore and ready to review. Rebased on top of ODH master and updated to Argo 2.12.1. I've added a tiny addition to the updater script, so now it can also copy over the upstream images to quay.io.

tumido commented:
Can somebody with the quay push permissions copy the Argo v2.12.1 images over to `opendatahub` repository, please?

tumido commented:
Tests are failing because the manifests are pointing to images which don't exist yet. It should be fixed when the images are copied over to `opendatahub` quay repo.

LaVLaS commented on odhargo/odhargo/base/default-sa-rolebinding.yaml:
```diff
@@ -0,0 +1,23 @@
+  name: argo-workflow-additions-role
+subjects:
+  - kind: ServiceAccount
+    name: default
```
This should be a ServiceAccount that is created by this component

tumido commented on odhargo/odhargo/base/default-sa-rolebinding.yaml:
```diff
@@ -0,0 +1,23 @@
+  name: argo-role
+subjects:
+  - kind: ServiceAccount
+    name: default
```
Sounds good. I'll change it then. :+1:

tumido commented:
Rebased

crobby commented:
Looks like the argo test failed.  Maybe the test needs updating for the new version?

crobby commented:
Also, a rebase will eliminate the aicoe-ci/build-check that errors out. It was just fixed today (Jan 8)

tumido commented:
> Looks like the argo test failed

@crobby that's because the images are missing. It probably got lost in the flood of comments, let me explain:
Upstream Argo uses DockerHub as their image hosting. That is not suitable for us due to DockerHub's rate limits, therefore [I've changed it](https://github.com/opendatahub-io/odh-manifests/pull/208/files#diff-d559981ab668d6e84e5e6283cd4949abc3513d2b96b1c81e38235de44d8fb9b6R22-R39) downstream, so the manifests fetch the images from `quay.io/opendatahub`. But the images were not copied over there yet. Can you or somebody with permissions to that Quay org create these repos and copy the images over there, please? (The copying can be done via the upgrade script provided in this PR, the repo creation has to be manual).

We'd need these 3 repos:
- `quay.io/opendatahub/workflow-controller` containing [this image](https://hub.docker.com/layers/argoproj/workflow-controller/v2.12.3/images/sha256-a875bb874c7020c9c89addf20144dc836e3383c13c728b1669ddfb7dcf8aded8?context=explore)
- `quay.io/opendatahub/argoexec` containing [this image](https://hub.docker.com/layers/argoproj/argoexec/v2.12.3/images/sha256-0069c4c4cdb15fdc695af1d2770936165b35afeecf9fa184bb3c6e43ae4bbb59?context=explore)
- `quay.io/opendatahub/argocli` containing [this image](https://hub.docker.com/layers/argoproj/argocli/v2.12.3/images/sha256-4bd0ce0c20110d581d9089e85efc7062dc3c57aa779779e5b0db5b840cbc09d5?context=explore)

crobby commented:
> > Looks like the argo test failed
> 
> @crobby that's because the images are missing. It probably got lost in the flood of comments, let me explain:
> Upstream Argo uses DockerHub as their image hosting. That is not suitable for us due to DockerHub's rate limits, therefore [I've changed it](https://github.com/opendatahub-io/odh-manifests/pull/208/files#diff-d559981ab668d6e84e5e6283cd4949abc3513d2b96b1c81e38235de44d8fb9b6R22-R39) downstream, so the manifests fetch the images from `quay.io/opendatahub`. But the images were not copied over there yet. Can you or somebody with permissions to that Quay org create these repos and copy the images over there, please? (The copying can be done via the upgrade script provided in this PR, the repo creation has to be manual).
> 
> We'd need these 3 repos:
> 
> * `quay.io/opendatahub/workflow-controller` containing [this image](https://hub.docker.com/layers/argoproj/workflow-controller/v2.12.3/images/sha256-a875bb874c7020c9c89addf20144dc836e3383c13c728b1669ddfb7dcf8aded8?context=explore)
> * `quay.io/opendatahub/argoexec` containing [this image](https://hub.docker.com/layers/argoproj/argoexec/v2.12.3/images/sha256-0069c4c4cdb15fdc695af1d2770936165b35afeecf9fa184bb3c6e43ae4bbb59?context=explore)
> * `quay.io/opendatahub/argocli` containing [this image](https://hub.docker.com/layers/argoproj/argocli/v2.12.3/images/sha256-4bd0ce0c20110d581d9089e85efc7062dc3c57aa779779e5b0db5b840cbc09d5?context=explore)

@tumido Thanks for clarifying things for me.  I have pushed those images to quay.io/opendatahub and made them public.  Let me know if there's anything else I can do to help out.

crobby commented:
/test all
Test after pushing argo images to quay.io

tumido commented:
Lol. I think the workflow test failed because the test workflow is pulling `docker/whalesay` image from DockerHub... :facepalm: I'll change the test workflow to use something from Quay.

tumido commented:
tests are flaky again 
![image](https://user-images.githubusercontent.com/7453394/104222786-106c3100-5443-11eb-9f80-09862c9534c1.png)

tumido commented:
/retest

tumido commented:
/retest

tumido commented:
Tests are still flaky, failures are unrelated to PR changes.

FTR, it seems upstream Argo is evaluating a move from DockerHub to Quay, so we may be able to use the upstream images, withouth the need to copy them over, next time. https://github.com/argoproj/argo/pull/4860

tumido commented:
/retest

nakfour commented on odhargo/README.md:
```diff
@@ -1,37 +1,64 @@
+
+1. Execute the [`gen_argo_manifests.sh`](gen_argo_manifests.sh) script within this folder.
+2. Latest manifests are fetched and this `README` is updated.
+3. Commit those changes.
```
@tumido no sure why we have a script to auto-update argo? Can you please explain the usecase for this. In our ODH installation we want to specify the release number because we test it.

tumido replied:
This is targeting the ODH maintainers, because with this script updating the ODH Argo component to be on par with upstream Argo would take a single command. In the PR description you can see what the script does:
1. It fetches the upstream manifests
2. Builds **all** the CRDs into a `cluster/base/crds.yaml` file
3. Copies over **alll** the cluster role aggregations to `cluster/base/cluster-roles.yaml`
4. Builds **all** the resources required for a **namespace install** into a `odhargo/base/namespace-install.yaml`, excluding the CRDs (so we maintain the `cluster` folder = requires cluster-admin permissions and `odhargo` folder = namespaced resources)

This way there's no need for manual copying of files anymore and all is kustomize driven based on the upstream manifests. The result manifests are stored in the same location every time. It changes the current structure of ODH Argo, sure. It makes it less error prone.

tumido replied:
It will also make the future change sets, when updating Argo to even newer versions, minimal. That's because it's a consistent way to update it. It also follows the upstream [Argo install guide](https://github.com/argoproj/argo/tree/master/manifests) as precisely as possible while maintaining the `odhargo/cluster` and `odhargo/odhargo` apps separation.

tumido replied:
Moved to https://github.com/opendatahub-io/odh-manifests/pull/289

nakfour commented on odhargo/cluster/base/cronworkflow-crd.yaml:
```diff
@@ -1,15 +0,0 @@
-    - cronwf
-    - cwf
-  scope: Namespaced
-  version: v1alpha1
```
Does the new version of argo not support cronworkflows?

tumido replied:
Yes it does, but it's relocated to [`cluster/base/crds.yaml`](https://github.com/opendatahub-io/odh-manifests/blob/2cc43411f49166dcb98ffb1b239d9c14c5efce39/odhargo/cluster/base/crds.yaml) since this file can be generated from the upstream manifests by the auto updater script and it would include all the CRDs at once, so we can't forget about any.

nakfour commented:
@tumido tested this and yes the dashboard link to the argo UI is broken, can we fix this in this PR? 
Also when launching the UI to argo server I got those two messages shown in the attached image. 
![Screen Shot 2021-01-13 at 4 12 39 PM](https://user-images.githubusercontent.com/18536575/104516703-67484680-55ba-11eb-9c61-2e71839a987d.png)

LaVLaS commented:
The argo-server pod is reporting errors due to not having cluster scope permissions to `cronworkflows`, `workflowtemplates`, `workflows`

```
time="2021-01-14T14:41:11.234Z" level=warning msg="finished unary call with code PermissionDenied" error="rpc error: code = PermissionDenied desc = cronworkflows.argoproj.io is forbidden: User \"system:serviceaccount:opendatahub:argo-server\" cannot list resource \"cronworkflows\" in API group \"argoproj.io\" at the cluster scope" grpc.code=PermissionDenied grpc.method=ListCronWorkflows grpc.service=cronworkflow.CronWorkflowService grpc.start_time="2021-01-14T14:41:11Z" grpc.time_ms=2.719 span.kind=server system=grpc
time="2021-01-14T14:41:11.234Z" level=warning msg="finished unary call with code PermissionDenied" error="rpc error: code = PermissionDenied desc = workflowtemplates.argoproj.io is forbidden: User \"system:serviceaccount:opendatahub:argo-server\" cannot list resource \"workflowtemplates\" in API group \"argoproj.io\" at the cluster scope" grpc.code=PermissionDenied grpc.method=ListWorkflowTemplates grpc.service=workflowtemplate.WorkflowTemplateService grpc.start_time="2021-01-14T14:41:11Z" grpc.time_ms=2.796 span.kind=server system=grpc
time="2021-01-14T14:41:11.331Z" level=warning msg="finished unary call with code PermissionDenied" error="rpc error: code = PermissionDenied desc = workflows.argoproj.io is forbidden: User \"system:serviceaccount:opendatahub:argo-server\" cannot list resource \"workflows\" in API group \"argoproj.io\" at the cluster scope" grpc.code=PermissionDenied grpc.method=ListWorkflows grpc.service=workflow.WorkflowService grpc.start_time="2021-01-14T14:41:11Z" grpc.time_ms=1.722 span.kind=server system=grpc
```

tumido commented:
@nakfour the error banners you see have the same cause as the error messages reported by @LaVLaS
![image](https://user-images.githubusercontent.com/7453394/104708620-3da83000-571e-11eb-9fe5-7946601a3c0a.png)

It seems like an upstream bug to look up resources at cluster scope if the deployments are limited to a namespace. I'll fix that upstream. In the current version we're gotta live with it. :/

It has no effect on Argo functionality since we use a namespaced installation.

tumido commented:
I've confirmed the errors reported are fixed upstream and will be part of the 2.13 release.

They don't effect functionality of Argo at all since they are caused by early loader on workflow filters by cronworkflow/workflowtemplate that is fired before the user is redirected into the namespace context. Once the user is redirected to the proper namespaced url, the loaders are fired again and only an annoying notifications is left. It's an UI bug, nothing serious.

Filters in question:
![image](https://user-images.githubusercontent.com/7453394/104734920-3a726b80-5741-11eb-9d2f-6e1183fe3976.png)

These errors don't appear anymore when using the `quay.io/argoproj/argocli:ui-v3` image. UI v3 will be part of Argo 2.13.

Tracking issue: https://github.com/argoproj/argo/issues/4885

crobby commented:
/test all

crobby commented:
I suspect that the tests will be happier after a rebase.  We have had a pair of enhancements to the jupyterhub tests this week that should make them more reliable.

LaVLaS commented:
The workflows worked for me with the new `argo-workflow` service account enabled.

LaVLaS reviewed (approved):
/lgtm
